odl/space/base_ntuples.py

```diff
@@ -403,6 +403,281 @@ def __repr__(self):
+            return NotImplemented
+
+        # Convert inputs that are ODL vectors to arrays so that the
+        # native Numpy ufunc is called later
+        inputs = tuple(
+            inp.asarray() if isinstance(inp, type(self)) else inp
+            for inp in inputs)
+
+        out = out1 = out2 = None
+        if len(out_tuple) == 1:
+            out = out_tuple[0]
+        elif len(out_tuple) == 2:
+            out1 = out_tuple[0]
+            out2 = out_tuple[1]
+
+        # Use some of the kwargs for `writable_array`
+        # TODO: propagate `order` when tensors are available
+        array_kwargs = {}
+        out_dtype = kwargs.get('dtype', None)
+        if out_dtype is not None:
+            array_kwargs['dtype'] = out_dtype
+
+        # Need new space for ufunc if dtype was changed
+        # TODO: use `shape` and `order` when tensors are available
+        if out_dtype is None or out_dtype == self.dtype:
+            out_space = self.space
+        else:
+            out_space = type(self.space)(self.size, out_dtype)
+
+        if method == '__call__':
+            if ufunc.nout == 1:
+                if out is None:
+                    out = out_space.element()
+                with writable_array(out, **array_kwargs) as out_arr:
+                    kwargs['out'] = (out_arr,)
+                    ufunc(*inputs, **kwargs)
+                return out
+
+            elif ufunc.nout == 2:
+                if out1 is None:
+                    out1 = out_space.element()
+                if out2 is None:
+                    out2 = out_space.element()
+                with writable_array(out1, **array_kwargs) as out1_arr:
+                    with writable_array(out2, **array_kwargs) as out2_arr:
+                        kwargs['out'] = (out1_arr, out2_arr)
+                        ufunc(*inputs, **kwargs)
+                return out1, out2
+
+            else:
+                raise NotImplementedError('nout = {} not supported'
+                                          ''.format(ufunc.nout))
+
+        else:
+            if out is None:
+                result = getattr(ufunc, method)(*inputs, **kwargs)
+                if np.isscalar(result):
+                    # This occurs for `reduce` with all axes
+                    return result
+                elif result is None:
+                    # Happens for in-place operations, currently only `at`
+                    return
+                else:
+                    # Wrap result in an appropriate space
+                    # TODO: use `shape` and `order` when tensors are available
+                    result_space = type(self.space)(result.size, result.dtype)
+                    return result_space.element(result.ravel())
+            else:
+                with writable_array(out) as out_arr:
+                    kwargs['out'] = (out_arr,)
+                getattr(ufunc, method)(*inputs, **kwargs)
+
+    # Some often used reductions (may not make sense in all cases)
```

Comment:
I'd actually say no since we want people to use the free functions whenever possible and we've seen some confusion regarding this before.